src/search/merge_and_shrink/fts_factory.cc

```diff
@@ -302,52 +307,72 @@ void FTSFactory::build_transitions_for_operator(OperatorProxy op) {
+        vector<int> &global_to_local_label =
+            transition_system_data_by_var[var_id].global_to_local_label;
+        vector<vector<Transition>> &local_label_to_transitions =
+            transition_system_data_by_var[var_id].local_label_to_transitions;
+        vector<LabelGroup> &local_to_global_labels =
+            transition_system_data_by_var[var_id].local_to_global_labels;
+        vector<int> &local_label_to_cost =
+            transition_system_data_by_var[var_id].local_label_to_cost;
+        assert(local_label_to_transitions.size() == local_to_global_labels.size());
+        assert(local_label_to_transitions.size() == local_label_to_cost.size());
         bool found_locally_equivalent_label_group = false;
-        for (size_t group_id = 0; group_id < existing_transitions_by_group_id.size(); ++group_id) {
-            const vector<Transition> &group_transitions = existing_transitions_by_group_id[group_id];
+        for (size_t local_label = 0; local_label < local_label_to_transitions.size(); ++local_label) {
+            const vector<Transition> &group_transitions = local_label_to_transitions[local_label];
             if (transitions == group_transitions) {
-                label_groups[group_id].push_back(label_no);
+                assert(global_to_local_label[label] == -1);
+                global_to_local_label[label] = local_label;
+                local_to_global_labels[local_label].push_back(label);
+                local_label_to_cost[local_label] = min(local_label_to_cost[local_label], label_cost);
                 found_locally_equivalent_label_group = true;
                 break;
             }
         }
 
         if (!found_locally_equivalent_label_group) {
-            existing_transitions_by_group_id.push_back(move(transitions));
-            label_groups.push_back({label_no});
+            int new_local_label = local_label_to_transitions.size();
+            local_label_to_transitions.push_back(move(transitions));
+            local_to_global_labels.push_back({label});
+            local_label_to_cost.push_back(label_cost);
+            assert(global_to_local_label[label] == -1);
+            global_to_local_label[label] = new_local_label;
         }
     }
 }
 
-void FTSFactory::build_transitions_for_irrelevant_ops(VariableProxy variable) {
-    int var_no = variable.get_id();
+void FTSFactory::build_transitions_for_irrelevant_ops(VariableProxy variable, const GlobalLabels &labels) {
+    int var_id = variable.get_id();
     int num_states = variable.get_domain_size();
-    int num_labels = task_proxy.get_operators().size();
 
     // Collect all irrelevant labels for this variable.
-    vector<int> irrelevant_labels;
-    for (int label_no = 0; label_no < num_labels; ++label_no) {
-        if (!is_relevant(var_no, label_no)) {
-            irrelevant_labels.push_back(label_no);
+    LabelGroup irrelevant_labels;
+    int cost = INF;
+    for (int label : labels) {
```

Comment:
rename to `global_label`?

Also, I think the code would be more readable if you separate the 2 loops. The loop to compute the cost could just loop over `irrelevant_labels` after the first loop is done. You could even use `std::min_element` instead of the second loop.

Reply:
Speaking of `std::min_element`: I recently found this and wasn't aware of it earlier. I think it could be used in a lot of places to simplify things but I find it a bit annoying that the call always gets so long because you pass in iterators. We could add overloads to the namespace `algorithm` that take a collection and just call `begin()` and `end()` on it.